src/Raven.Server/Utils/Metrics/MeterMetric.cs

```diff
@@ -131,42 +104,150 @@ public void Mark()
-            var oneSec = TimeSpan.FromSeconds(1);
-            for (int i = index; i >= 0; i--)
+                return result;
+
+            var nowNs = _now();
+            var nowUtc = DateTime.UtcNow;
+            var bucketSeconds = BucketDurationNanoseconds / (double)Clock.NanosecondsInSecond;
+
+            var buckets = new List<(long Start, double Rate)>();
+            for (int i = 0; i < BucketCount; i++)
             {
-                var d = _m15Rate[i];
-                if (filterEmpty == false || Math.Abs(d) > double.Epsilon)
-                    results[now.ToString()] = Math.Round(d, 1);
-                now = now - oneSec;
+                ref var bucket = ref _buckets[i];
+                var start = Volatile.Read(ref bucket.Start);
+                if (start == 0)
+                    continue;
+
+                if (nowNs - start > MaxWindowNanoseconds)
+                    continue;
+
+                var count = Volatile.Read(ref bucket.Count);
+                if (filterEmpty && count == 0)
+                    continue;
+
+                buckets.Add((start, count / bucketSeconds));
             }
-            for (int i = _m15Rate.Length - 1; i >= 0; i--)
+
+            buckets.Sort((a, b) => a.Start.CompareTo(b.Start));
+
+            // Emit raw 100ms samples so diagnostics can plot the exact workload shape (no smoothing).
+            var raw = new DynamicJsonValue();
+            foreach (var (start, rate) in buckets)
             {
-                var d = _m15Rate[i];
-                if (Math.Abs(d) > double.Epsilon)
-                    results[now.Add(TimeSpan.FromSeconds(-i)).ToString()] = Math.Round(d, 1);
-                now = now - oneSec;
+                var ageSeconds = (nowNs - start) / (double)Clock.NanosecondsInSecond;
+                var timestamp = nowUtc - TimeSpan.FromSeconds(ageSeconds);
+                raw[timestamp.ToString("O", CultureInfo.InvariantCulture)] = Math.Round(rate, 1);
             }
-            return r;
+
+            result["Raw"] = raw;
+            return result;
+        }
+
+        private void AddToBucket(long timestamp, long value, long duration)
+        {
+            var quantum = timestamp / BucketDurationNanoseconds;
+            var bucketStart = quantum * BucketDurationNanoseconds;
+            var index = (int)(quantum % BucketCount);
+
+            ref var bucket = ref _buckets[index];
+            var start = Volatile.Read(ref bucket.Start);
+            if (start != bucketStart)
+            {
+                if (bucketStart > start)
+                {
+                    // We have moved forward in time: clear the previous contents and stamp this bucket with
+                    // the new quantum so readers can distinguish fresh samples from stale data after wrap-around.
+                    Interlocked.Exchange(ref bucket.Count, 0);
+                    Interlocked.Exchange(ref bucket.DurationSum, 0);
```

Comment:
No compares, why Interlocked needed here? Could you not write just 0s and have release semantics ensured by the `Volatile.Write` below?

Reply:
Yes, we could technically do `Volatile.Write` here. Let me reason about it a little bit more, because I think there are a few potential rough edges still.